From ef6e97f78a11c9114dceb4c394b3c07824db5eee Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Wed, 22 Sep 2021 19:49:47 +0300
Subject: [PATCH] [SW-55215] gaudi2: condition sp sram scrub

The SP SRAM content needs to be preserved over reset, so do not scrub
it each reset.
Do this only once per power cycle.

Change-Id: Icb81ebcc62a4595a23b967956968886b42c5196f
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index bed2d38..e083d44 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -22,6 +22,7 @@
 #include <stdint.h>
 #include <hl_boot_if.h>	/* requires stdint.h */
 #include <gaudi2/gaudi2_reg_map.h>
+#include <gaudi2/gaudi2_fw_if.h>
 #include <common/pci_ids.h>
 
 #include <bitops.h>
@@ -79,7 +80,7 @@
  * ECC errors are observed due to byte copy to SP
  * memory which turns to be read-modify-write transaction
  */
-void preboot_sp_sram_init(void)
+static void gaudi2_sp_sram_init(void)
 {
 	u32 init_sts;
 	u64 start;
@@ -103,6 +104,21 @@
 	}
 }
 
+void preboot_sp_sram_init(void)
+{
+	struct gaudi2_cold_rst_data cold_rst_data;
+
+	cold_rst_data.data = regmap_readl(mmCOLD_RST_DATA);
+	if (!cold_rst_data.spsram_init_done) {
+		gaudi2_sp_sram_init();
+
+		cold_rst_data.spsram_init_done = 1;
+		regmap_writel(cold_rst_data.data, mmCOLD_RST_DATA);
+	}
+
+	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_SP_SRAM_EN, true);
+}
+
 void hl_plat_early_init(void)
 {
 	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_FW_IATU_CONF_EN, true);
